File: app/api/v1/subscriptions.py
"""Subscription & Billing endpoints."""
from typing import Optional
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from app.core.auth import UserContext, require_permission, get_current_user
from app.services.subscription_service import SubscriptionService

router = APIRouter(prefix="/subscriptions", tags=["Subscriptions"])
_svc = SubscriptionService()


# ── Request Schemas ──

class SubscribeIn(BaseModel):
    plan_slug: str  # starter | growth | pro


class UpgradeIn(BaseModel):
    new_plan_slug: str


class DowngradeIn(BaseModel):
    new_plan_slug: str


class PurchaseAddonIn(BaseModel):
    addon_slug: str
    quantity: int = 1
    shipping_address: Optional[dict] = None


class AdminUpdatePlanIn(BaseModel):
    name: Optional[str] = None
    price: Optional[float] = None
    monthly_price: Optional[float] = None
    description: Optional[str] = None
    features: Optional[list] = None
    limits: Optional[list] = None
    not_included: Optional[list] = None
    highlight: Optional[bool] = None
    highlight_label: Optional[str] = None
    cta_text: Optional[str] = None
    discount_label: Optional[str] = None
    razorpay_plan_id: Optional[str] = None
    is_active: Optional[bool] = None


# ── Public Endpoints ──

@router.get("/plans")
async def list_plans():
    """List all available subscription plans (public)."""
    return await _svc.get_plans()


@router.get("/addons")
async def list_addons():
    """List all available add-on products (public)."""
    return await _svc.list_addons()


# ── User Endpoints ──

@router.get("/status")
async def check_subscription(
    user: UserContext = Depends(require_permission("subscription.read")),
):
    """Check if user has an active subscription."""
    return await _svc.check_active(user.user_id)


@router.get("")
async def get_subscription(
    user: UserContext = Depends(require_permission("subscription.read")),
):
    """Get full subscription details."""
    return await _svc.get_subscription(user)


@router.post("/verify")
async def verify_subscription(
    user: UserContext = Depends(get_current_user),
):
    """Verify subscription status with plan details."""
    return await _svc.verify_subscription(user)


@router.post("/free-trial")
async def start_free_trial(
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """Start a 14-day free trial (one per user)."""
    return await _svc.start_free_trial(user)


@router.post("/subscribe")
async def subscribe(
    body: SubscribeIn,
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """
    Create a subscription. Returns Razorpay subscription details
    including short_url for checkout.
    """
    return await _svc.subscribe(user, body.plan_slug)


@router.post("/cancel")
async def cancel_subscription(
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """
    Cancel subscription. Access continues until current billing period ends.
    Trials are cancelled immediately.
    """
    return await _svc.cancel_subscription(user)


@router.post("/upgrade")
async def upgrade_plan(
    body: UpgradeIn,
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """
    Upgrade to a higher plan. Takes effect immediately.
    Returns Razorpay checkout link for the new plan.
    """
    return await _svc.upgrade_plan(user, body.new_plan_slug)


@router.post("/downgrade")
async def downgrade_plan(
    body: DowngradeIn,
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """
    Schedule a downgrade. Takes effect at next billing cycle.
    """
    return await _svc.downgrade_plan(user, body.new_plan_slug)


@router.post("/addons/purchase")
async def purchase_addon(
    body: PurchaseAddonIn,
    user: UserContext = Depends(require_permission("subscription.write")),
):
    """
    Purchase an add-on (e.g., printer). Returns Razorpay order
    for one-time payment checkout.
    """
    return await _svc.purchase_addon(
        user, body.addon_slug, body.quantity, body.shipping_address
    )


# ── Admin Endpoints ──

@router.get("/admin/list")
async def admin_list_subscriptions(
    status: Optional[str] = Query(None),
    limit: int = Query(50, le=200),
    offset: int = Query(0, ge=0),
    user: UserContext = Depends(require_permission("subscription.admin")),
):
    """Admin: View all subscriptions, optionally filtered by status."""
    return await _svc.admin_list_subscriptions(status, limit, offset)


@router.patch("/admin/plans/{plan_id}")
async def admin_update_plan(
    plan_id: int,
    body: AdminUpdatePlanIn,
    user: UserContext = Depends(require_permission("subscription.admin")),
):
    """Admin: Update plan pricing, features, or Razorpay plan ID."""
    return await _svc.admin_update_plan(plan_id, body.model_dump(exclude_unset=True))

File: frontend/src/api/subscriptions.js
/**
 * Subscription API hooks for Bittu frontend.
 * Uses the backend at VITE_API_URL.
 */

const API_BASE = import.meta.env.VITE_API_URL || "https://api.merabittu.com";

function getHeaders(token) {
  return {
    "Content-Type": "application/json",
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
  };
}

async function handleResponse(res) {
  const data = await res.json();
  if (!res.ok) {
    const error = new Error(data.detail || "Request failed");
    error.status = res.status;
    error.code = data.error_code;
    throw error;
  }
  return data;
}

// ─── Public (no auth) ────────────────────────────────────────

export async function fetchPlans() {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/plans`);
  return handleResponse(res);
}

export async function fetchAddons() {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/addons`);
  return handleResponse(res);
}

// ─── Authenticated ───────────────────────────────────────────

export async function getSubscription(token) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions`, {
    headers: getHeaders(token),
  });
  return handleResponse(res);
}

export async function getSubscriptionStatus(token) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/status`, {
    headers: getHeaders(token),
  });
  return handleResponse(res);
}

export async function verifySubscription(token) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/verify`, {
    method: "POST",
    headers: getHeaders(token),
  });
  return handleResponse(res);
}

export async function startFreeTrial(token) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/free-trial`, {
    method: "POST",
    headers: getHeaders(token),
  });
  return handleResponse(res);
}

export async function subscribe(token, planSlug) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/subscribe`, {
    method: "POST",
    headers: getHeaders(token),
    body: JSON.stringify({ plan_slug: planSlug }),
  });
  return handleResponse(res);
}

export async function cancelSubscription(token) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/cancel`, {
    method: "POST",
    headers: getHeaders(token),
  });
  return handleResponse(res);
}

export async function upgradePlan(token, newPlanSlug) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/upgrade`, {
    method: "POST",
    headers: getHeaders(token),
    body: JSON.stringify({ new_plan_slug: newPlanSlug }),
  });
  return handleResponse(res);
}

export async function downgradePlan(token, newPlanSlug) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/downgrade`, {
    method: "POST",
    headers: getHeaders(token),
    body: JSON.stringify({ new_plan_slug: newPlanSlug }),
  });
  return handleResponse(res);
}

export async function purchaseAddon(token, addonSlug, quantity = 1, shippingAddress = null) {
  const res = await fetch(`${API_BASE}/api/v1/subscriptions/addons/purchase`, {
    method: "POST",
    headers: getHeaders(token),
    body: JSON.stringify({
      addon_slug: addonSlug,
      quantity,
      shipping_address: shippingAddress,
    }),
  });
  return handleResponse(res);
}
